### ISSUE TITLE
Maxwell Run

### ISSUE THREAD

**Issue by Gidemsky** [zendesk/maxwell#1629]
Hey,

I use maxwell to ready MySQL binlog and send those events to RabbitMQ. Everything works well except the fact, I have several events that get the **save timestamp** due to MySql contains only a 32bit timestamp field.
I read there are two possible solutions to add millisecond timestamps for each event:

- Embed a millisecond-precise date field in the data
- Write a custom producer add a fake millisecond precision to the output row when the row is sent to the RabbitMQ

The second solution works better for me. The only thing I need to know, is whether Maxwell can guarantee the order I send the row is the same order they have been written at the MySql binlog. Also If it does, is there any simple way to add this fake millisecond timestamp to the row?

An example of my need looks like that (**the field milli_ts added to the header**):
```
mysql> insert into test.e set m = 4.2341, c = now(3), comment = 'I am a creature of light.';
{
   "database":"test",
   "table":"e",
   "type":"insert",
   "ts":1477053217,
   "milli_ts":123456789123454667,
   "xid":23396,
   "commit":true,
   "position":"master.000006:800911",
   "server_id":23042,
   "thread_id":108,
   "primary_key": [1, "2016-10-21 05:33:37.523000"],
   "primary_key_columns": ["id", "c"],
   "data":{
      "id":1,
      "m":4.2341,
      "c":"2016-10-21 05:33:37.523000",
      "comment":"I am a creature of light."
   }
}
```

**Comment by osheroff** [comment]
> The only thing I need to know, is whether Maxwell can guarantee the order I send the row is the same order they have been written at the MySql binlog. 

yes, with the redis producer that's guaranteed. 

> Also If it does, is there any simple way to add this fake millisecond timestamp to the row?

check out `RowMap#putExtraAttribute` -- it should get you what you need.

**Comment by Gidemsky** [comment]
Thank you for your attention.

> yes, with the redis producer that's guaranteed.

So that means, the fact I use RabbitMQ producer, doesn't guarantee that the order of the binlog rows will be sent with the same order to RabbitMQ?

**Comment by osheroff** [comment]
>So that means, the fact I use RabbitMQ producer, doesn't guarantee that the order of the binlog rows will be sent with the same order to RabbitMQ?

Oh, I'm sorry, I mis-read.  Any producer that doesn't partition its data (rabbit, redis, file) should have the same guarantee

**Comment by Gidemsky** [comment]
Ok, sounds perfect.

Basically, the reason I need to create a millisecond timestamp feature is to guarantee using the same binlog's rows order with Maxwell's JSON result. Today, we have an edge case caused by the same timestamp for several binlog's rows, although those rows weren't really been created at the same time, but at the same second.
For example, let's assume we have 3 MySQL queries that were executed one after the other:
```
insert into test.e set m = 4.2341, c = now(3), comment = 'I am a creature of light.';
update test.e set m = 5.444, c = now(3) where id = 1;
update test.e set m = 123456, c = now(3) where id = 1;
```
In such a case Maxwell's result I going to use will be:
```
{
   "database":"test",
   "table":"e",
   "type":"insert",
   "ts":1477053217,
   "xid":23396,
   "commit":true,
   "position":"master.000006:800911",
   "server_id":23042,
   "thread_id":108,
   "primary_key": [1, "2016-10-21 05:33:37.523000"],
   "primary_key_columns": ["id", "c"],
   "data":{
      "id":1,
      "m":4.2341,
      "c":"2016-10-21 05:33:37.523000",
      "comment":"I am a creature of light."
   }
}
```
```
{
   "database":"test",
   "table":"e",
   "type":"update",
   "ts":1477053217,
   ...
   "data":{
      "id":1,
      "m":5.444,
      "c":"2016-10-21 05:33:54.631000",
      "comment":"I am a creature of light."
   },
   "old":{
      "m":4.2341,
      "c":"2016-10-21 05:33:37.523000"
   }
}
```
```
{
   "database":"test",
   "table":"e",
   "type":"update",
   "ts":1477053217,
   ...
   "data":{
      "id":1,
      "m":123456,
      "c":"2016-10-21 05:33:54.631000",
      "comment":"I am a creature of light."
   },
   "old":{
      "m":5.444,
   }
}
```
If some of those JSON order goes wrong, I can't know what is the right query order, due to all the MySQL queries executed the same second and has got the same timestamp.

Is it possible, there is any simpler configuration solution, could help me using or sign those Maxwell's JSON results with the binlog's order without creating a custom producer? Maybe the _output_binlog_position_ field in the config.properties file?

In case I still have to create my custom producer, do I need to use the Maxell source code or I can use a different approach?

**Comment by osheroff** [comment]
> Is it possible, there is any simpler configuration solution, could help me using or sign those Maxwell's JSON results with the binlog's order without creating a custom producer? Maybe the output_binlog_position field in the config.properties file?

you may want to look into the output of `--output_xoffset` -- this gives you an auto-increment field within a transaction (xid): 
something like `sort_by([row.ts, row.xid, row.xoffset])` should give you a sort order in like 99.999999% of cases.  fwiw I actually kind of like your hacky fake-millisecond approach (might be actually worth implementing behind a flag).  We'd have to do it as microsecond precision to support more than 1k rows per second but millisecond precision has been oft-requested, although I don't know how many use cases are sort-order and how many use cases want real data that we can never retrieve.  Might be worth floating an "enhancement" issue and see how many emojis it receives....

> In case I still have to create my custom producer, do I need to use the Maxell source code or I can use a different approach?

you create a standalone java project for your custom producer that consumes the maxwell jar as a maven artifact.  after you build your artifact, drop it into `lib` in the directory of the untarred maxwell release.

**Comment by Gidemsky** [comment]
Cool, that sounds good enough for my needs.

The last part which doesn't clear for me is where can I see the output_xoffset's record.
I've signed this as `output_xoffset=true` in the config.properties file, but couldn't find any record that has been added to the JSON or Maxwell's MySQL databases.

**Comment by osheroff** [comment]
yeah, for some reason this was coded to only appear on row that aren't the "last row in the transaction".  I have no idea why this is. 

I'm looking at this more, though, and I don't actually think my recommendation of using [ts, xid, xoffset] is correct for your use case; `ts` refers to a timestamp in the mysql binlog when the row was modified, not when the transaction was committed; so you could have out-of-order operations if you relied on it.

what about if we just coded up a microsecond-precision field and called it like `push_ts` or `maxwell_ts` or something -- it can indicate the microsecond at which maxwell generated the row for sending to the producer. 

We can put a little hack in there to make sure that no two rows share the same microsecond.

**Comment by Gidemsky** [comment]
I think If maxwell could produce a microsecond-precision that would solve a lot of troubles for us. Do you think you can do it soon?

but also, let say I have the same timestamp, why using the xid field, doesn't solve the problem? Basically, I talk about the case I have the same timestamp.

**Comment by osheroff** [comment]
@Gidemsky because the `ts` field itself is no good as a sort key; because it records the timestamp when the row was modified (via sql), not when the transaction was committed, if you sort based on `ts` you can end up re-ordering operations.

**Comment by osheroff** [comment]
https://github.com/zendesk/maxwell/pull/1639

**Comment by osheroff** [comment]
fixed in 1.30.0
